### DISCUSSION TITLE
Brute Force IP ban possibilities

### DISCUSSION THREAD

**Discussion by woytekbode** [dani-garcia/vaultwarden#2162] [Ideas]
I have just started experimenting with Vaultwarden. Thank you for all the work that you guys are doing! I came across the possibility of hosting my own vaultwarden instance via Home Assistant while searching for better password managers.

I'm running a RPi4 with HA OS so vaultwarden runs as an addon in Home Assistant. Therefore it's not possible for me to run fail2ban as I see recommended here and there. HA has its own options for setting a maximum number of login tries, would something like that be possible for vaultwarden as well?

Of course if I have misunderstood and there are alternatives available I am all ears!

**Comment by BlackDex** [comment]
We try to stay clear from adding those kind of options since they are more suited tools like some reverse proxies which have this (not sure which one from the top of my head). 

Also, not sure if that is possible, depending if you are able to set some `CAP_` options you could use: https://github.com/crazy-max/docker-fail2ban

**Comment by dani-garcia** [comment]
Lately I've been thinking that maybe we should include a very basic rate limiting option in vaultwarden, the same way we expose and document Rocket's TLS option, yet we recommend a reverse proxy anyway.

I'm not so much worried about logins, which are kind of automatically ratelimited by the fact that hashing takes time, but by other requests like two factor or the admin page.

Even just a basic max requests per second per IP for login, two factor and admin seems enough for me, and with some sensible defaults we could improve the security for all the users who don't enable fail2ban or similar (which is probably a lot as it's not very easy to configure).

Using something like https://docs.rs/governor/0.3.2/governor/index.html should be fairly straightforward.

What do you think?

**Reply by BlackDex** [reply]
That could work i think. Depending if you want to have this all in memory or also store those IP's, and add exceptions or maybe even a /admin interface to manage it, like remove specific blocked/limited IP's it could be getting a bit bigger then just a simple rate-limit per IP.

Also, if we are going to switch to async being the main branch, it has much better TLS support, and it should work better out-of-the box, so a reverse proxy is maybe not fully needed, though i probably would still recommend it, but it does provide an argument for having something like this build-in.

**Reply by BlackDex** [reply]
Though, we still need to have a reverse proxy for people who want to have the web-sockets.

**Reply by dani-garcia** [reply]
I've implemented a very basic rate limiting here: https://github.com/dani-garcia/vaultwarden/pull/2165. Tell me what you think.

**Reply by woytekbode** [reply]
Wow did not expect to see action in this so quickly! I agree that a simple rate limit would be a very basic but effective additional security measure. I briefly looked at the code and I think it looks good! Can't say much about the specifics, not experienced with the code or rust...

**Reply by ChenyuWang98** [reply]
I added fail2ban protection. Testing is in progress to brute force the admin. But before fail2ban took effect, vaultwarden took the lead in restricting logins.` [vaultwarden::ratelimit][ERROR] Too many admin requests`. I now want to know how to remove this restriction.I searched for relevant information on google and only saw this discussion.

**Reply by BlackDex** [reply]
See
https://github.com/dani-garcia/vaultwarden/blob/48836501bf348386d9bb1378fb56db33c19d3732/.env.template#L354

**Reply by ChenyuWang98** [reply]
Thank you so much

**Reply by GodjeNl** [reply]
As far as I can see, if vaultwarden is hosted behind a proxy (e.g. nginx or cloudflare). This will block IP of the upstream proxy-server and therefore all IP's. Is there a way to prevent this?

**Reply by kpfleming** [reply]
If that's the case, the blocking should be done upstream of Vaultwarden; there's no reason to allow the connection through the proxy at all.

**Comment by GodjeNl** [comment]
> If that's the case, the blocking should be done upstream of Vaultwarden; there's no reason to allow the connection through the proxy at all.

A reason could be to enable HTTPS which is the recommended method according to https://github.com/dani-garcia/vaultwarden/wiki/Enabling-HTTPS . But to be able  to handle the blocking upstream, Vaultwarden should write the original IP (X-Real-IP or X-Forwarded-For) to the log.  The upstream server cannot possibly know the submitted password is incorrect.

**Reply by BlackDex** [reply]
We can do that, just provide the correct header we need to use.

https://github.com/dani-garcia/vaultwarden/blob/897bdf8343148ee4e4f1a937cbc501ad3f743b4a/.env.template#L290-L292

**Comment by arpanghosh8453** [comment]
For Cloudflare Tunnels (zero trust proxy), the `IP_HEADER=X-Forwarded-For` works as .env variable. You need to make sure the `.env` variables are not being overridden by the `config.json`. You can verify this from the admin panel.
